### PR TITLE
✨ Nhj public output

### PR DESCRIPTION
**Pull request by NHJohnson**
## Description

This is create public and protected consensus outputs per #1045 and also incorporates some of the flexible-input stuff from #1021, although it does not address filtering expressions. 

Fixes https://github.com/d3b-center/bixu-tracker/issues/1045

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- 
## How Has This Been Tested?

One question relates to the retained info input, which is used by both the vcf2maf and kfdrc_vcf2maf_public steps. Do these in fact use the same retained info, or are two inputs needed, one for each step? 

- [X] On Cavatica: https://cavatica.sbgenomics.com/u/nathanj/test-project/tasks/8624221d-33eb-4245-8def-7c1f8ef934a4/#

**Test Configuration**:
* Environment:
* Test files:

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [x] I have committed any related changes to the PR

### CODE REVIEW

File: workflow/kfdrc_consensus_calling.cwl
Comment (sickler-alex) [inline workflow/kfdrc_consensus_calling.cwl:31]: These lines can be deleted.

**Review by sickler-alex — CHANGES_REQUESTED** [review]
Looks good, just one minor change.

File: workflow/kfdrc_consensus_calling.cwl
Comment (migbro) [inline workflow/kfdrc_consensus_calling.cwl:25]: it seems you are missing `filter_expression`, which is needed to properly add soft filters to the vcf. Otherwise, nothing happens during the public filter portion. Could rename this variable, since we have the `bcftools_public_filter` from `filter_names` to something like `soft_filter_names` and `soft_filter_expression`. I might want to do that myself at some point to add clarity to the annotation workflow...

**Review by migbro — CHANGES_REQUESTED** [review]
Mostly good except `filter_expression` is missing, so the public outputs come out no difference from the protected

**Comment by migbro** [comment]
I also noticed that the expected behavior, of a default output of using a min of two callers didn't happen. It seemed fine at the script level I thought...regardless, an `ncaller` param should/could be added with a default values of 2 that is passed to the consensus step. That should resolve it for sure, but might want to check the script too to make sure that
```python
parser.add_argument('--ncallers', type=int, default=2,
                        help='Number of callers required for consensus [2]')
```
is being interpreted correctly

File: workflow/kfdrc_consensus_calling.cwl
Comment (NHJohnson) [reply workflow/kfdrc_consensus_calling.cwl:31]: ✅

File: workflow/kfdrc_consensus_calling.cwl
Comment (NHJohnson) [reply workflow/kfdrc_consensus_calling.cwl:25]: Are you referring to the argument for which the defaults are 
[ "vc.getGenotype(" + inputs.input_normal_name + ").getDP() <= 7", "AF > 0.001" ]
?

These are the ones for which I was having trouble getting it to work because of the normal_name dependency. On this branch, those filters are still set in the variant_filter step, but there is no ability to change the defaults. I will fix this on my other branch as soon as I figure out how if that's ok.

Comment (migbro) [reply workflow/kfdrc_consensus_calling.cwl:25]: Ah, yes, I missed that. Which points me to the real bug, I see this error in the log for that filter step:
`2021-04-15T19:50:46.598474932Z sh: 1: input_normal_name: not found`
resulting from this arg (copied from the cmd.log): ` --filter-expression "vc.getGenotype($(input_normal_name)).getDP() <= 7"` not being resolved properly

**Comment by NHJohnson** [comment]
@migbro I will add `ncallers` as an option to this workflow, but it looks like the apparent failure of the default may be due to every single locus being marked as a hotspot, which is kind of concerning on its own. I will try to understand if this is an error. I thought something like this happened before, because the annotation in the single-caller VCFs was read by pysam as a tuple rather than an int, etc., but we had fixed it.

**Comment by migbro** [comment]
> @migbro I will add `ncallers` as an option to this workflow, but it looks like the apparent failure of the default may be due to every single locus being marked as a hotspot, which is kind of concerning on its own. I will try to understand if this is an error. I thought something like this happened before, because the annotation in the single-caller VCFs was read by pysam as a tuple rather than an int, etc., but we had fixed it.

Hmmm, yeah I remember that. Maybe check to make sure the docker is pointing at the updated version?

**Comment by NHJohnson** [comment]
It was not pointing to the updated version because! I forgot to merge it after you guys approved it like a month ago. Now waiting for it to be merged in the SBG registry.

**Comment by NHJohnson** [comment]
Ok, I believe this one works:
https://cavatica.sbgenomics.com/u/nathanj/test-project/tasks/e2b85185-dc87-45a6-96c0-500a6ed65bf7/#

filter_expressions are functional with defaults but do not offer desired flexibility; will address in future PR after I investigate how this can be achieved.

**Comment by NHJohnson** [comment]
Well, actually, is the depth filter still not being applied? The output doesn't seem to contain any depth filtered variants, but I don't see the problem in job.json for that step ...

**Comment by NHJohnson** [comment]
And the right stuff seems to be in the VCF headers ...

**Comment by NHJohnson** [comment]
Tumor / normal names were wrong, held over from a previous test. Trying again ...

**Comment by migbro** [comment]
Catching up - if you still don't see depth filter hits, you can test with a lower threshold. It could just be that the resulting consensus file ended up with variants that pass that filter just fine.

**Comment by NHJohnson** [comment]
I'm afraid I'm not sure. I ran again with the fixed tumor / normal names:
https://cavatica.sbgenomics.com/u/nathanj/test-project/tasks/aba92302-b3c9-45c4-814b-b5b7be7a2e86/

but the DP filter still doesn't seem to be applied to any variants in the output. I believe there are variants that should be filtered this way, as if I do a
`bcftools view -i "DP[0]<=7"`
on the output I get ~500 hits

**Comment by NHJohnson** [comment]
Ok, that thing with the single quotes seems to have fixed it. 
https://cavatica.sbgenomics.com/u/nathanj/test-project/tasks/ffc14fa0-de46-49d2-89e8-d0861a43cbe6/#

**Comment by migbro** [comment]
Yeah, I think it looks better. Just for my sanity, can you run on the inputs from here:
https://cavatica.sbgenomics.com/u/nathanj/test-project/files/#q?path=NORM_INPUT_TEST
The inputs you used were from that run where our hotspots file was too hot. These are from the much smaller hotspot inputs that we intend to use. Thanks!

**Comment by NHJohnson** [comment]
This seems to have worked:
https://cavatica.sbgenomics.com/u/nathanj/test-project/tasks/d4afd08e-6a4a-43b0-8570-e7c22a8cfb9e/

The number of hotspots is much smaller, but they still exist.

**Comment by NHJohnson** [comment]
HotSpotAllele issue addressed in consensus step by #106

**Review by sickler-alex — APPROVED** [review]

**Review by migbro — APPROVED** [review]
I ran one more test run here and was pretty happy with it: https://cavatica.sbgenomics.com/u/nathanj/test-project/tasks/d5f469dd-fc9a-4680-bdcd-ca081ea01e04/
As you mentioned, the hotspot designation will be updated from another place

**Pull request merged**
